File: OOP-and-Functional-Programming_Samples/OOP/Simulated_Assembly_instructions_and_Samples/java/Programa.java
import java.util.Stack;

public class Programa{

    private Instruccion[] instrucciones;
    private int contador;
    private int pc;
    private Stack<Integer> pila;
    
    public Programa(){
        this.instrucciones = new Instruccion[10]; //Ponemos una capacidad inicial de 10 instrucciones, por ejemplo.
        this.contador = 0;
        this.pc = 0;
        this.pila = new Stack<>();
    }
    
    //Añadimos instrucciones al vector dinámico. Si el vector está lleno, duplicamos su tamaño.
    public void anyadir(Instruccion instr){
        if (contador == instrucciones.length){
            Instruccion[] nuevoArray = new Instruccion[instrucciones.length * 2];
            for (int i = 0; i < instrucciones.length; i++){
                nuevoArray[i] = instrucciones[i];
            }
            instrucciones = nuevoArray;
        }
        instrucciones[contador++] = instr;
    }
    
    //Listamos las instrucciones del programa con su índice.
    public void listar(){
        for (int i = 0; i < contador; i++){
            System.out.println(i + ": " + instrucciones[i].mostrar());
        }
    }
    
    //Ejecutar el programa.
    public void ejecutar(){
        pila.clear();
        pc = 0;
        while (pc < contador){
            instrucciones[pc].ejecutar(this);
        }
    }
    
    //Incrementar pc.
    public void inc(){
        pc++;
    }
    
    //Cambiar valor de pc para dar un salto.
    public void salto(int nuevoPc){
        pc = nuevoPc;
    }

    public Stack<Integer> getPila(){
        return pila;
    }

}

File: OOP-and-Functional-Programming_Samples/OOP/Simulated_Assembly_instructions_and_Samples/c++/Programa.cc
#include "Programa.h"
using namespace std;

Programa::Programa(){
    capacidad = 10;
    instrucciones = new Instruccion*[capacidad];
    contador = 0;
    pc = 0;
}

//Destructor.
Programa::~Programa(){
    for (int i = 0; i < contador; i++){
        delete instrucciones[i];
    }
    delete[] instrucciones;
}

//Añadimos instrucciones al vector dinámico. Si el vector está lleno, duplicamos su tamaño.
void Programa::anyadir(Instruccion* instr){
    if (contador == capacidad){
        int nuevaCapacidad = capacidad * 2;
        Instruccion** nuevoArray = new Instruccion*[nuevaCapacidad];
        for (int i = 0; i < capacidad; i++){
            nuevoArray[i] = instrucciones[i];
        }
        delete[] instrucciones;
        instrucciones = nuevoArray;
        capacidad = nuevaCapacidad;
    }
    instrucciones[contador++] = instr;
}

//Listamos las instrucciones del programa con su índice.
void Programa::listar() const{
    for (int i = 0; i < contador; i++){
        cout << i << ": " << instrucciones[i]->mostrar() << endl;
    }
}

//Ejecutar el programa.
void Programa::ejecutar(){
    while (!pila.empty()){
        pila.pop();
    }
    pc = 0;

    while (pc < contador){
        instrucciones[pc]->ejecutar(this);
    }
}

//Incrementar pc.
void Programa::inc(){
    pc++;
}

//Cambiar valor de pc para dar un salto.
void Programa::salto(int nuevoPc){
    pc = nuevoPc;
}

stack<int>& Programa::getPila(){
    return pila;
}
